### PR TITLE
fix(custom-resource): set physical resource ID as soon as possible to prevent failure in delete event

### LINKED ISSUE #2870: Delete validation records upon `copilot env init` failures
### Background
If an application is associated with a domain, (e.g. by running `copilot app init  --name my-app --domain my-domain.com`), when we provision am environment, we request a certificate for the environment domain, (e.g. `test.my-app.my-domain.com`), and upsert the validation record to the hosted zone.

### The Problem
When the certificate failed to validate, for example, because of a timeout, CloudFormation always fail to delete the environment stack.

### Why
When the certificate failed to validate, we do not fill the `PhysicalResourceID` with the certificate ARN. This results in this [conditional](https://github.com/aws/copilot-cli/blob/cbf43493e8e42d3d04c2394349199016229eaa3c/cf-custom-resources/lib/dns-cert-validator.js#L596) evaluating to `false` during a `DELETE` event (triggered because of the CloudFormation rollback), hence skipping the handler.

We should make sure to remove any upserted record during a `DELETE` event.

### PR DESCRIPTION
**Pull request by Lou1415926**
Fix #2870 and #2882

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: cf-custom-resources/lib/dns-cert-validator.js [hunk 1]
```diff
@@ -78,43 +82,49 @@ let report = function (
   });
 };
 
+const aliasesChanged = async function (
```
Comment (efekarakus) [inline cf-custom-resources/lib/dns-cert-validator.js]: nit: I think this fn doesn't need to be `async`

File: cf-custom-resources/lib/dns-cert-validator.js [hunk 2]
```diff
@@ -78,43 +82,49 @@ let report = function (
...
+  let prevAliases = await getAllAliases(
+      oldResourceProperties.Aliases
+  );
+  let aliasesToDelete = [...prevAliases,].filter(function (itm) {
+    return !aliases.has(itm);
+  });
+  let aliasesToAdd = [...aliases,].filter(function (itm) {
+    return !prevAliases.has(itm);
+  });
+  return !aliasesToAdd.length + aliasesToDelete.length === 0;
```
Comment (efekarakus) [inline cf-custom-resources/lib/dns-cert-validator.js]: 😮  is this doing the right thing vs:
```suggestion
  return !((aliasesToAdd.length + aliasesToDelete.length) === 0);
```

File: cf-custom-resources/lib/dns-cert-validator.js [hunk 3]
```diff
@@ -164,30 +188,52 @@ const requestCertificate = async function (
...
   );
 
   await acm
     .waitFor("certificateValidated", {
       // Wait up to 9 minutes and 30 seconds
       $waiter: {
-        delay: 30,
-        maxAttempts: 19,
+        delay: 1,
+        maxAttempts: 1,
```
Comment (efekarakus) [inline cf-custom-resources/lib/dns-cert-validator.js]: Do we want these changes?

File: cf-custom-resources/lib/dns-cert-validator.js [hunk 4]
```diff
@@ -545,50 +593,51 @@ exports.certificateRequestHandler = async function (event, context) {
...
-          props.EnvHostedZoneId,
-          props.RootDNSRole,
-          props.Region
+          sansToUse,
+          acm
         );
-        responseData.Arn = physicalResourceId = certificateArn;
+        responseData.Arn = physicalResourceId = response.CertificateArn;  // Set physicalResourceId as soon as we can.
+        options = await waitForValidationOptionsToBeReady(response.CertificateArn, sansToUse, acm);
+        await validateCertificate(options, envRoute53, appRoute53, props.EnvHostedZoneId, response.CertificateArn, acm);
```
Comment (efekarakus) [inline cf-custom-resources/lib/dns-cert-validator.js:621]: oh i see, so now we separated the certificate request logic compared to before where we would merge all three operations into one place

File: cf-custom-resources/lib/dns-cert-validator.js [hunk 5]
```diff
@@ -545,50 +593,51 @@ exports.certificateRequestHandler = async function (event, context) {
...
-            return !aliases.has(itm);
-          });
-          var aliasesToAdd = [...aliases].filter(function (itm) {
-            return !prevAliases.has(itm);
-          });
-          if (aliasesToAdd.length + aliasesToDelete.length === 0) {
-            break;
-          }
+        let shouldUpdate = await aliasesChanged(event.OldResourceProperties, aliases);
+        if (!shouldUpdate) {
```
Comment (efekarakus) [inline cf-custom-resources/lib/dns-cert-validator.js]: nit:
```suggestion
        if (!aliasesChanged(event.OldResourceProperties, aliases)) {
```

File: cf-custom-resources/lib/dns-cert-validator.js (hunk 3 above)
Comment (Lou1415926) [reply cf-custom-resources/lib/dns-cert-validator.js]: oh no.. good catch and sorry :woman-facepalming:

File: cf-custom-resources/lib/dns-cert-validator.js (hunk 2 above)
Comment (iamhopaul123) [reply cf-custom-resources/lib/dns-cert-validator.js]: Maybe just
```
return aliasesToAdd.length + aliasesToDelete.length !== 0;
```

File: cf-custom-resources/lib/dns-cert-validator.js [hunk 6]
```diff
@@ -78,43 +82,49 @@ let report = function (
...
+  let prevAliases = getAllAliases(
+      oldResourceProperties.Aliases
+  );
+  let aliasesToDelete = [...prevAliases,].filter(function (itm) {
+    return !aliases.has(itm);
+  });
+  let aliasesToAdd = [...aliases,].filter(function (itm) {
+    return !prevAliases.has(itm);
+  });
+  return aliasesToAdd.length + aliasesToDelete.length !== 0;
```
Comment (efekarakus) [inline cf-custom-resources/lib/dns-cert-validator.js]: im dumb why not just 
```suggestion
  return aliasesToAdd.length + aliasesToDelete.length;
```

**Review by efekarakus — APPROVED** [review]

File: cf-custom-resources/lib/dns-cert-validator.js [hunk 7]
```diff
@@ -478,13 +525,15 @@ const getAllAliases = function (aliases) {
   } catch (error) {
     throw new Error(`Cannot parse ${aliases} into JSON format.`);
   }
-  var aliasList = [];
-  for (var m in obj) {
-    aliasList.push(...obj[m]);
+  let aliasList = [];
+
+  for (const [_, aliases] of Object.entries(obj)) {
```
Comment (iamhopaul123) [inline cf-custom-resources/lib/dns-cert-validator.js:530]: just curious why are we making this change 🤔

File: cf-custom-resources/lib/dns-cert-validator.js (hunk 6 above)
Comment (Lou1415926) [reply cf-custom-resources/lib/dns-cert-validator.js]: :homer-disappearing:

**Review by iamhopaul123 — APPROVED** [review]

File: cf-custom-resources/lib/dns-cert-validator.js (hunk 7 above)
Comment (Lou1415926) [reply cf-custom-resources/lib/dns-cert-validator.js:530]: Not much tbh. Just b/c it has slightly more readability +  JSLint always complains with the message "The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype."

**Pull request merged**